### PR TITLE
gh-104504: cases generator: Add `--warn-unreachable` to the mypy config

### PR DESCRIPTION
**Pull request by AlexWaygood**
Backport `typing.assert_never` to the `cases_generator` directory, so that mypy doesn't complain about the fact that we're using a feature new in Python 3.11 when we have `--python-version 3.10` in our config file.

* Issue: gh-104504

### CODE REVIEW

File: Tools/cases_generator/generate_cases.py [hunk 1]
```diff
@@ -171,7 +172,6 @@ def effect_str(effects: list[StackEffect]) -> str:
                 popped, pushed = stacking.get_stack_effect_info_for_macro(instr)
             case parsing.Pseudo():
                 instr = self.pseudo_instrs[thing.name]
-                popped = pushed = None
```
Comment (AlexWaygood) [inline Tools/cases_generator/generate_cases.py:174]: Mypy was eagerly narrowing the type of `pushed` here to be `None`, meaning that it was inferring the `else` branch on line 188 to be unreachable. Assigning to `popped` and `pushed` at the same point where their types are declared solves this problem.

File: Tools/cases_generator/generate_cases.py [hunk 2]
```diff
@@ -388,7 +388,6 @@ def write_metadata(self, metadata_filename: str, pymetadata_filename: str) -> No
                 case parsing.Macro():
                     format = self.macro_instrs[thing.name].instr_fmt
                 case parsing.Pseudo():
-                    format = None
```
Comment (AlexWaygood) [inline Tools/cases_generator/generate_cases.py:391]: Similarly, here mypy was eagerly narrowing the type of `format` to `None`, meaning that it was inferring the `else` branch on line 396 to be unreachable.

File: Tools/cases_generator/generate_cases.py (hunk 1 above)
Comment (corona10) [reply Tools/cases_generator/generate_cases.py:174]: I think that you can revert the L186 code into the original one that I modified in the previous PR.

````diff
@@ -182,8 +182,7 @@ def effect_str(effects: list[StackEffect]) -> str:
                     assert target_instr
                     target_popped = effect_str(target_instr.input_effects)
                     target_pushed = effect_str(target_instr.output_effects)
-                    if pushed is None:
-                        assert popped is None
+                    if popped is None and pushed is None:
                         popped, pushed = target_popped, target_pushed
                     else:
                         assert popped == target_popped
````

Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:174]: > Assigning to `popped` and `pushed` at the same point where their types are declared solves this problem.

Still, that feels unsatisfactory. Each case is supposed to set these variables, so the initialization at the declaration point ought to be unnecessary. It seems mypy is just wrong here?

Even though I agree that this is probably the best solution for now, in general I feel that there's a lot of churn in this piece of code due to limitations in mypy. :-(

**Review by gvanrossum — APPROVED** [review]

File: Tools/cases_generator/generate_cases.py (hunk 1 above)
Comment (AlexWaygood) [reply Tools/cases_generator/generate_cases.py:174]: > Still, that feels unsatisfactory. Each case is supposed to set these variables, so the initialization at the declaration point ought to be unnecessary

Right, this PR reduces symmetry between the branches in a less-than-ideal way. I didn't see that at first (apologies, I'm new to this code).

I think there's an alternative solution, but I might not get to it for a day or two; I'm traveling this weekend.

**Comment by AlexWaygood** [comment]
@gvanrossum, are you happier with the new refactor? Now `pushed` and `popped` are always `str`, never `None`, in `get_stack_effect_info()`. Similarly, and `format` is always `str`, never `None`, in `write_metadata()`. It resolves the mypy reachability-analysis issue just as well as my previous idea, and restores the symmetry between the branches that we have on `main`.

**Comment by AlexWaygood** [comment]
(The test_concurrent_futures failure on Azure Pipelines is unrelated)

File: Tools/cases_generator/generate_cases.py [hunk 3]
```diff
@@ -171,25 +170,23 @@ def effect_str(effects: list[StackEffect]) -> str:
                 popped, pushed = stacking.get_stack_effect_info_for_macro(instr)
             case parsing.Pseudo():
                 instr = self.pseudo_instrs[thing.name]
-                popped = pushed = None
                 # Calculate stack effect, and check that it's the the same
                 # for all targets.
-                for target in self.pseudos[thing.name].targets:
+                for idx, target in enumerate(self.pseudos[thing.name].targets):
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:175]: Clever idea. I wrote a whole paragraph arguing against using enumerate() here, then realized it wasn't worth arguing over. :-)

File: Tools/cases_generator/generate_cases.py (hunk 2 above)
Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:391]: Still feels like a mypy bug to me, but your fix is fine.

**Review by gvanrossum — APPROVED** [review]
Let's check this in.

**Comment by AlexWaygood** [comment]
Thanks for the review!

File: Tools/cases_generator/generate_cases.py (hunk 3 above)
Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:175]: FYI, the saga continues. Currently we get errors from pyright on this: "popped is possibly unbound" (on all occurrences of popped and pushed in this function below here).

File: Tools/cases_generator/generate_cases.py [hunk 4]
```diff
@@ -388,17 +383,15 @@ def write_metadata(self, metadata_filename: str, pymetadata_filename: str) -> No
...
-                        if format is None:
+                        if idx == 0:
                             format = target_instr.instr_fmt
                         else:
                             assert format == target_instr.instr_fmt
-                    assert format is not None
                 case _:
-                    typing.assert_never(thing)
+                    assert_never(thing)
             all_formats.add(format)
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:395]: And ditto here for `format`.

File: Tools/cases_generator/generate_cases.py (hunk 3 above)
Comment (AlexWaygood) [reply Tools/cases_generator/generate_cases.py:175]: > FYI, the saga continues. Currently we get errors from pyright on this: "popped is possibly unbound" (on all occurrences of popped and pushed in this function below here).

Yeah, and mypy will also give you that error if you add `--enable-error-code=possibly-undefined` to the mypy config :/

There is a third way of fixing this. I'll make a PR and see how you like it.

**Pull request merged**